### PR TITLE
types: added http header types to reply

### PR DESCRIPTION
**Pull request by skwee357**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/reply.d.ts [hunk 1]
```diff
@@ -14,6 +14,8 @@ export interface ReplyGenericInterface {
...
...equest-headers' | 'access-control-request-method' | 'age' | 'allow' | 'authorization' | 'cache-control' | 'cdn-cache-control' | 'connection' | 'content-disposition' | 'content-encoding' | 'content-language' | 'content-length' | 'content-location' | 'content-range' | 'content-security-policy' | 'content-security-policy-report-only' | 'cookie' | 'dnt' | 'date' | 'etag' | 'expect' | 'expires' | 'forwarded' | 'from' | 'host' | 'if-match' | 'if-modified-since' | 'if-none-match' | 'if-range' | 'if-unmodified-since' | 'last-modified' | 'link' | 'location' | 'max-forwards' | 'origin' | 'prgama' | 'proxy-authenticate' | 'proxy-authorization' | 'public-key-pins' | 'public-key-pins-report-only' | 'range' | 'referer' | 'referrer-policy' | 'refresh' | 'retry-after' | 'sec-websocket-accept' | 'sec-websocket-extensions' | 'sec-websocket-key' | 'sec-websocket-protocol' | 'sec-websocket-version' | 'server' | 'set-cookie' | 'strict-transport-security' | 'te' | 'trailer' | 'transfer-encoding' | 'user-agent' | 'upgrade' | 'upgrade-insecure-requests' | 'vary' | 'via' | 'warning' | 'www-authenticate' | 'x-content-type-options' | 'x-dns-prefetch-control' | 'x-frame-options' | 'x-xss-protection' | string;
```
Comment (Eomm) [inline types/reply.d.ts:17]: Do we have some automation for this list?

> public-key-pins

First time in my life I read it, but:

> in 2017 browsers deprecated and removed HPKP support

https://en.wikipedia.org/wiki/HTTP_Public_Key_Pinning

I'm afraid we will get PRs to add whatever HTTP header in the world

**Review by mcollina — CHANGES_REQUESTED** [review]
Custom http headers are very common. This should also allow any string.

**Comment by Uzlopak** [comment]
Does this even work????????

type name = 'matteo' | 'eomm' | string

Is afaik because of type widening the same as 
type name = string 

So i doubt that this has any benefit.

**Review by Uzlopak — CHANGES_REQUESTED** [review]
Blocking because type widening makes this pointless.

**Comment by mcollina** [comment]
> Blocking because type widening makes this pointless.

This is incorrect. It improves autocompletion.

**Comment by Uzlopak** [comment]
@mcollina 

Please check yourself

https://www.typescriptlang.org/play?#code/C4TwDgpgBAdghgWwgZygXigcgXYwID2mUAPloQgsWcsAE4CWMA5gLABQHAxgTLVBAAeiMABsIALliIU6LJg5A

no autocompletion.

**Comment by Eomm** [comment]
For reference:

image

**Comment by skwee357** [comment]
A lot of comments, I'll try to address them all:

* I took the list of headers from Rust ```http``` package (as I was porting some Rust code to Node) → https://github.com/hyperium/http/blob/master/src/header/name.rs
* My PR does allow for arbitrary header name, as it allows ```string```
* The motivation behind my PR, is to improve auto-complete and avoid stupid mistakes like ```contest-location``` instead of ```content-location```. I understand it's not 100% bulletproof, but fastify does provide types for request headers → https://github.com/fastify/fastify/blob/main/types/request.d.ts#L59

I'm open for suggestions on how to improve it, but I think typing is necessary.

**Comment by Uzlopak** [comment]
I take vscode (which uses monaco like typescript playground) as reference. And in vscode this change has no effect.

**Comment by sergburn** [comment]
Fastify's header name constants come from [`@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e532df3be90a6f1399387c9e6e93c7da276d55e8/types/node/http.d.ts#L49C75-L49C75), so maybe this is the way they should be defined properly, because as Eomm shows, code completion doesn't work this way.
Code completion would look like this:

image


However. Node types do not define response header names, maybe for a reason...

**Comment by skwee357** [comment]
@sergburn thanks. I agree that it's better to take the header names from ```@types/node```. I wonder why they don't define reply headers, though.

**Comment by skwee357** [comment]
Started a discussion in ```@types/node```. Let's see what they will answer and proceed based on their response https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66753

**Comment by DemonHa** [comment]
if we do something like this:
```ts
type headers = 'mateo' | 'eonm' | (string & {});
```
it will give us the autocompletion but also we can define general strings
image

@Semigradsky @Eomm @mcollina

**Comment by mcollina** [comment]
`type headers = 'mateo' | 'eonm' | (string & {});` is supported syntax and it works.

Good job.

@skwee357 could you update the PR?

**Comment by Uzlopak** [comment]
Nice trick. Is there an explaination?

**Comment by mcollina** [comment]
https://x.com/matteocollina/status/1706595817409093698?s=20

**Comment by climba03003** [comment]
Reading through the threads, it may suffer from another issue.
https://github.com/microsoft/TypeScript/issues/29729#issuecomment-554927299

There will be problems on object index.

**Comment by metcoder95** [comment]
> if we do something like this:
> 
> ```ts
> type headers = 'mateo' | 'eonm' | (string & {});
> ```
> 
> it will give us the autocompletion but also we can define general strings image


This past week has been full of new learnings. First `split` with the extra arg, and now this 😅 🪄

**Comment by Uzlopak** [comment]
I patched it now according to @DemonHa suggestion.

Now the question is, is it safe to merge @climba03003 or do you recommend a hard no?

**Comment by DemonHa** [comment]
> Nice trick. Is there an explaination?

@Uzlopak 
I don't know a clear explanation on this. Might be interesting to look at [this](https://tsplay.dev/mMjZZW). If you hover on type X looks something like this: 

```ts
{
    [x: string & {}]: number;
    alert: number;
    tree: number;
    treegrid: number;
    treeitem: number;
}
```

while if you hover over Y:
```ts
{
    [x: string]: number;
}
```

Somehow it prevent the union to be squashed

[Stackoverflow](https://stackoverflow.com/questions/69793164/typescript-weird-type-intersection-of-string)

**Comment by DemonHa** [comment]
> Nice trick. Is there an explaination?

@Uzlopak 
There is a great explanation [here](https://stackoverflow.com/questions/75262513/what-is-the-purpose-of-string)

**Comment by skwee357** [comment]
> `type headers = 'mateo' | 'eonm' | (string & {});` is supported syntax and it works.
> 
> Good job.
> 
> @skwee357 could you update the PR?

I'm working on adding the types to `@types/node` instead, see here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66783

Once it will be merged, I'll update this PR to use the types from that package, rather that having fastify maintain its own types

**Comment by Uzlopak** [comment]
@DemonHa 
@skwee357 
@mcollina 
@Eomm 
@metcoder95 
@Semigradsky 
@climba03003 

PTAL

File: types/utils.d.ts [hunk 2]
```diff
@@ -78,3 +78,12 @@ export type RecordKeysToLowercase<Input> = Input extends Record<string, unknown>
...
   : Input;
+
+type OmitIndexSignature<T> = {
+  [K in keyof T as string extends K ? never : number extends K ? never : K]: T[K];
+};
+
+/**
+ * HTTP header strings
+ */
+export type HttpHeader = keyof OmitIndexSignature<http.OutgoingHttpHeaders> | (string & {}); // eslint-disable-line @typescript-eslint/ban-types
```
Comment (climba03003) [inline types/utils.d.ts:89]: Please add a comment to warn for the usage on return value.
This hack is perfectly fine for `input` value but not for `return` value.
https://github.com/microsoft/TypeScript/issues/29729#issuecomment-554927299

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Uzlopak** [comment]
I followed the link provided by @climba03003 and saw that we can do [Record<never, never>](https://github.com/microsoft/TypeScript/issues/29729#issuecomment-1548233969) instead of `{}`. So to avoid that unnessary eslint ignore, i used that.

**Review by climba03003 — APPROVED** [review]

**Pull request merged**